### ISSUE TITLE
 ctdFindProfiles() should handle lon/lat if they are columns in the data slot

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#2026]
This came up in debugging for #2024 and #2025

**Comment by dankelley** [comment]
Oh, hang on.  I don't think it should be looking in the `data` slot.  It ought to be looking in the `metadata` slot, since lon and lat are not actually measured by the CTD. 

Put another way, the problem is that my as-yet-not-pushed code in the `rsk` branch is storing lon and lat in the wrong place!  I'll move that and retest.

**Comment by dankelley** [comment]
I will need to alter `ctdFindProfile()` a bit to solve this, because presently it is only working on things stored in the `data` slot.  What I will do is to see if `longitude` and `latitude` are in the `metadata` slot, and if they are the same length as `data$pressure` ... and then I'll subset in the same way as I subset the items in `data`.

**Comment by dankelley** [comment]
I'm starting to look at the `ctdFindProfile()` code and I see that I will need also to adjust the `ctdTrim()` code, because that's what does the focussing.

I want to think about this for a while, though.  I don't quite like an action that might have a false positive. I don't see (at this moment) why `metadata$longitude` would be of length greater than 1 and *not* be of the same length as `data$pressure`, but maybe I'm just not thinking this through well.  The only reason for making this odd decision is my desire to store only "measured" things in `data`.  But we can already have inferred things there, e.g. a user might very insert some quantity of interest, for convenience.  And the CTD doesn't measure salinity, yet we put it in the `data` slot, don't we?

Hm... I am leaning to making `read.rsk()` put lon and lat in the `data` slot, which doesn't require changing old functions like `ctdTrim()` and `ctdFindProfiles()` in ways that would not be compatible with their old action.

I plan to decide on this today because I don't like waiting a week to decide on 3 lines of code.  I may have a meeting with AH and will ask for a second opinion then.

**Comment by dankelley** [comment]
Subject to my *ad movere* intention, I decided to add the trim-within-metadata scheme, modifying `ctdTrim()` and `ctdFindProfiles()` within "develop" branch, commit dfe569fbdf81d388c3206b994713c0c0668afa2b.

**Comment by Brendan-Higgins** [comment]
I was able to build oce from the .rsk branch and see that there is now location information in the metadata slot when I call the read.rsk() and ctdFindProfilesRBR() functions. Working from the same example .rsk file we've been referring to, it looks like each cast has a few hundred locations (all of which are pretty close to each other). Which specific coordinates for each cast are the plot functions using?

**Comment by dankelley** [comment]
Can you submit code so we know what you are doing?  (That is, what do you mean by 'plot functions'?)

It depends on how you are plotting.  There are multiple ways. As profiles.  As a section with x axis being time.  As a scan-plot.  Etc.

I think in the first two of these cases it uses the mean location during the time during which the CTD was doing a particular profile.  However, I would need to check on that, because I wrote the code a while ago and then had a wild flurry of grading to do.  Please provide some sample code so I know which of the various possible plot types you mean.  (Refer to your data file as `sample.rsk` or something -- use whatever name I've been using.)

Really, the answer should be in the docs, but I have not yet documented most of how things are working with this new RSK functionality, because the  code is not finalized.  Amongst the things that are required to finalize the code are:

1. Is there a document to which I can refer that discusses how the device is connected to the phone/tablet/laptop/whatever?  Ideally that doc will be citable, with a URL, but sometimes we just cite manufacturer's documents with title, date etc and no URL.
2. What happens if the user presses a button (or whatever) to indicate the start of a profile?  I think @richardsc suggested that this is possible, and so if the info on those presses is in the rsk file, *and* if no profile can be made without pressing that button, then we should likely use that time.  But I don't want to get into a guessing-game where maybe it got pressed for a given profile and maybe not ... then the code would have to switch between two methods.  And maybe somebody took some time to have lunch before they remembered to press the button etc.  Again, this really comes down to what an RBR manual says the procedure is.  Maybe the RBR won't store data until a button is pressed?
3. @richardsc has suggested that something is wrong with my method of connecting phone/tablet/laptop/whatever times to profile times.  Perhaps we can discuss that sometime this month.

**Comment by Brendan-Higgins** [comment]
I'm trying to make plots with distance on the x-axis:

library(oce)

f <- "C:/Users/Brendan/Desktop/copy of late ctd/209642_20220808_1504.rsk"

# Read data.  The time offset is particular to this data file.
d <- read.rsk(f, allTables=TRUE, tzOffsetLocation=-8)

# Chop into profiles, using informatoin in RBR headers
ctds0 <- ctdFindProfilesRBR(as.ctd(d))

# Ignore top 2m and bottom 3m, which may be incorrect.
ctds1 <- lapply(ctds0,
                function(ctd) {
                  p <- ctd[["pressure"]]
                  subset(ctd, 2 < p & p < max(p) - 3)
                })

#create a section with a subset of the ctd casts
section_aialik = list(ctds1[[16]],ctds1[[17]],ctds1[[18]],
                      ctds1[[19]],ctds1[[20]],ctds1[[21]]) %>%
  as.section()

#make contour plot with distance on the x-axis
plot(section_aialik, xtype='distance', ztype='image')

**Comment by Brendan-Higgins** [comment]
I'm trying to make a section plot with distance on the x-axis:

`library(oce)

f <- "C:/Users/Brendan/Desktop/copy of late ctd/209642_20220808_1504.rsk"

# Read data.  The time offset is particular to this data file.
d <- read.rsk(f, allTables=TRUE, tzOffsetLocation=-8)

# Chop into profiles, using informatoin in RBR headers
ctds0 <- ctdFindProfilesRBR(as.ctd(d))

# Ignore top 2m and bottom 3m, which may be incorrect.
ctds1 <- lapply(ctds0,
                function(ctd) {
                  p <- ctd[["pressure"]]
                  subset(ctd, 2 < p & p < max(p) - 3)
                })

#create a section with a subset of the ctd casts
section_aialik = list(ctds1[[16]],ctds1[[17]],ctds1[[18]],
                      ctds1[[19]],ctds1[[20]],ctds1[[21]]) %>%
  as.section()

#make contour plot with distance on the x-axis
plot(section_aialik, xtype='distance', ztype='image')
`

**Comment by Brendan-Higgins** [comment]
1.  The RBR user guide for 'standard loggers' explains the process in the geotagging section on page 24
 [https://rbr-global.com/products/software/mobile-software/](url)

2.  Pressing the button in the app adds a geotag to the gpx file but the data is collected whether the user presses the button or not. And location data is seemingly collected  either way as well. I think the table:

ctd[[1]]@metadata[["geodata"]][["origin"]]

may show if each point was made automatically through the general logging process or manually by pressing the button. This is a guess on my part but I think it makes sense.

Additional info:
There are multiple ways to initiate data collection on the instrument. A user can either start/stop recording through the app, use a physical switch on the instrument, or rely on a sensor on the instrument that detects if it's in the water or not. I don't know how RBR uses these various methods in regards to cast detection

**Comment by dankelley** [comment]
@Brendan-Higgins please note that I edited your comment so that the code starts with <pre>```R</pre>
and ends with 
<pre>```</pre>.

Doing that makes the code show up properly, so please use that convention to save me from having to do it so I can see what you mean.

Regarding https://github.com/dankelley/oce/issues/2026#issuecomment-1366950703 can you let me know if you are using the github "develop" branch of oce?  That is where I'll make changes, if changes are required.  It's also what I'll refer to, if we need to discuss the coding in detail.

Regarding https://github.com/dankelley/oce/issues/2026#issuecomment-1366957376, thanks for the info.  I think @richardsc and I will likely meet sometime in January to make some decisions on detailed things.  But I can certainly address your actual question from https://github.com/dankelley/oce/issues/2026#issuecomment-1366950703 tomorrow morning (it's night here now).

Thanks!  Dan.

**Comment by dankelley** [comment]
I had a look at the manual, and it's quite impressive, how they have set things up with external devices.  I downloaded the software and I think it should help me to learn about the file contents.  The software isn't bad, actually, although I am disappointed to report that it crashes (on macOS) when I click the icon to see the maps ... which is obviously my main concern.

Anyway, I'm going to fiddle with this for a while this morning.  My present plan is to try two things. 

1. If possible infer locations from "tagged" spots (if that's the right word) if they exist, and take those spots to be the profile locations.
2. If tagged points don't exist (or maybe if they are a long time ago) instead find RBR location by interpolating to phone location using time as index.  (This is what I do at the moment.)

Both schemes have problems, I think.  In scheme 1, what if the user didn't press the "tag" button on the phone?  In scheme 2, what if the operator started sampling on the dock, but then the RBR was used for a day or so, sampling in local waters, while the phone travelled with the operator who might have driven from the dock to the town to get lunch, or maybe driven an hour to a larger town to buy some gear?

Frankly, though, we can't make oce into a magic box.  I think it ought to do something reasonable that might be helpful in most instances.  If the setup was weird (in the ways listed in the previous paragraph, or in other ways, then it's up to the user to figure things out.  The vessel will always have a GPS and a sensible system would be to have a logbook in which somebody wrote down things like "Station 2, middle of channel, LON=..., LAT=..., time=..." like in the old days.

@richardsc and I might like to touch base at some point to discuss the things I brought up in this email.  Meantime, I'm going to look in the rsk file to see whether I can find the tagged points (and whether they match up with the location markers I see in the RSK software on my laptop).  Then I'll need to decide what to do about using scheme 1, scheme 2, or -- more likely -- a scheme that uses method 1 if possible and then switches to method 2 for some profiles, if nobody pressed that "tag" button for a given profile.

PS. I am aware that "tag" might not be the right word.

**Comment by dankelley** [comment]
@Brendan-Higgins Please let me know if you don't want anybody seeing the station locations ... if so I'll delete this and any related comment that reveals this information.  (My hope is that you permit display of the map here because otherwise you, @richardsc and I will need to keep track of not just github issues but also private emails.)

I think we have 39 location data (from the below).  By comparison, the table `d@metadata$breaksdata` lists 28 `start` and `end` entries.  I take that to mean that Ruskin has identified 28 profiles.  This confuses me because when I was last working on this I thought we had 23 stations.  (I am not worried that 39 does not equal either 28 or 23, since the manual gives me the impression that the instrument will add in extra fixes if the phone gets a better ones, after the user presses a button and before the descent starts.)

@Brendan-Higgins can I get a definitive statement on how many stations you think there were?  (I apologize in probably going back to the beginning, but this is likely something you know off the top of your head, and finding such a statement in emails and github issues is hard.)

If you click on the Details line below, you'll see some code I wrote, the textual output, and a graph.  I think the `origin=="auto"` cases are what we want to examine.  I don't know what the `origin=="manual"` cases mean.  (Maybe a button got pressed just 3 times, when somebody wrote down a vessel-based GPS measurement as well, to calibrate/test?)


<details>

**rsk_with_location_20221229a.R**

```R
# This requires the latest "rsk" branch from github, and a renamed local file.

library(oce)
data(coastlineWorldFine, package="ocedata")
f <- "ctd_with_location.rsk"
# NB. the time offset is particular to this data file.
d <- read.rsk(f, allTables=TRUE, tzOffsetLocation=-8, debug=0)
geodata <- d[["geodata"]]
str(geodata)
print(table(geodata$origin))
lon <- geodata$longitude
lat <- geodata$latitude
origin <- geodata$origin
asp <- 1 / cos(mean(range(lat))*pi/180)
plot(lon[origin=="auto"], lat[origin=="auto"], asp=asp,
xlab="Longitude", ylab="Latitude")
points(lon[origin=="manual"], lat[origin=="manual"], col=2, pch=20)
polygon(coastlineWorldFine[["longitude"]], coastlineWorldFine[["latitude"]],
    col=rgb(210/255, 180/255, 140/255, alpha=0.1))

```

**Output text from above**

```R
'data.frame':	15491 obs. of  6 variables:
 $ tstamp      :integer64 1659999898080 1660000206077 1660000192000 1660000202000 1660000212000 1660000222000 1660000231000 1660000242000 ... 
 $ latitude    : num  59.9 59.9 59.9 59.9 59.9 ...
 $ longitude   : num  -150 -150 -150 -150 -150 ...
 $ accuracy    : num  4.55 4.55 6.07 9.1 4.55 ...
 $ accuracyType: chr  "" "" "" "" ...
 $ origin      : chr  "auto" "auto" "gpx" "gpx" ...

  auto    gpx manual 
    39  15449      3 
```

**Plot from above**

![a](https://user-images.githubusercontent.com/99469/209954282-b8a80255-d9e3-4a0b-9fbe-5b6caa797297.png)

**Comment by dankelley** [comment]
@Brendan-Higgins with respect to your question https://github.com/dankelley/oce/issues/2026#issuecomment-1366324175 about lon and lat, a section plot will use the first longitude and latitude for a given profile.  

But the results may change a little because I think I should be using just the `origin=="auto"` indices of the lon/lat data for reference, not the whole lon/lat dataset.  I am assuming that the `"auto"` values have been assessed to be more accurate than the other values.  This is just guesswork, but if I were writing the RBR software, I'd pay special attention to the times when users pressed a button on their phones, e.g. waiting a longer time to get a signal before storing it.  This is the sort of thing that @richardsc and I might want to chat about.  It's not good to just make guesses on things, without at least getting a second opinion.

**Comment by dankelley** [comment]
I decided to compare two methods for inferring lon and lat. Rather than write a paragraph explaining, the code is as below (really, this is just for me and @richardsc; the code is in `R/rsk.R` of "rsk" branch commit 87fe34c3c71fa13cf2e63766a45ea80cf8818cc9).
```R
        if (!is.null(geodata)) {
            geodata$time <- numberAsPOSIXct(geodata$tstamp/1e3,type="unix")+tzOffsetLocation*3600
            look <- geodata$origin == "auto"
            res@metadata$latitude <- approx(geodata$time, geodata$latitude, res@data$time)$y
            res@metadata$latitudeNew <- approx(geodata$time[look], geodata$latitude[look], res@data$time)$y
            res@metadata$longitude <- approx(geodata$time, geodata$longitude, res@data$time)$y
            res@metadata$longitudeNew <- approx(geodata$time[look], geodata$longitude[look], res@data$time)$y
            message("lon-lat may be wrong; see https://github.com/dankelley/oce/issues/2024#issuecomment-1345373099")
            message("TEST: examine both longitude and longitudeNew etc")
        }
 ```
And I test this with
```R
png("A.png")
par(mfrow=c(2, 1), mar=c(3, 3, 2, 1), mgp=c(2, 0.7, 0))
hist(d[["longitude"]]-d[["longitudeNew"]], breaks=100)
hist(d[["latitude"]]-d[["latitudeNew"]], breaks=100)
dev.off()
```
yielding as below.  I'll focus on lat because that's easy to interpret i.t.o. distance.  The diff between old and new methods is typically under 0.001 deg, which is 100 m.  The RBR software reports GPS uncertainties of 40m or so in the cases I've examined, so this difference is not large. However, I do think it is better to use the "new" method and I am inclined to switch to that, unless either @Brendan-Higgins or @richardsc objects before Saturday.

I am conscious of the facts that (a) @Brendan-Higgins doesn't start work until I am finished for the day and that (b) @richardsc is likely having a bit of time off.  Therefore, I do not plan on doing much more on this (apart from the aforementioned change in lon/lat) until @richardsc and I can meet over zoom or in person, perhaps next week or some other time in Jan.

![A](https://user-images.githubusercontent.com/99469/209964866-b7c6982b-9f1b-41f0-bb67-7ff811158d54.png)

**Comment by dankelley** [comment]
I have decided to focus on those "auto" data.  Rather than describe what I've done here, I have done so in the docs, which now read as follows (snippets from `?read.rsk`).  All of this is kind of confusing, because the below (and a lot of the above) has drifted from the title I originally used.  This often happens when an issue leads to code changes that lead to other issues.

<img width="745" alt="Screenshot 2022-12-30 at 1 00 31 PM" src="https://user-images.githubusercontent.com/99469/210094982-c8f0d560-0673-4db3-b268-c010b62cf126.png">

...

<img width="761" alt="Screenshot 2022-12-30 at 1 00 41 PM" src="https://user-images.githubusercontent.com/99469/210095006-bda98f43-408b-4d91-bf0f-8b75c3106a5a.png">

**Comment by dankelley** [comment]
Actually, now that I run some of @Brendan-Higgins' code, I see that there is sense in putting the longitude and latitude into the `data` slot because then the following will leave the output `length(metadata$longitude)` equal to `length(data$pressure)`, which we would want if we say used `plotTS()` on the objects.

The alternative would be to make `subset` look at `length(metadata$longitude)` to see if  it equals `data$pressure` but where do we stop with such new things?  Latitude belongs there.  But maybe time also.  Maybe some other things that I just cannot think of at the moment.  Also, the action of `subset,ctd-method` has been centred on the `data` slot for over a decade, and has been documented to do so, and this is a strong argument against changing it.



```R
ctds1 <- lapply(ctds0,
                function(ctd) {
                  p <- ctd[["pressure"]]
                  subset(ctd, 2 < p & p < max(p) - 3)
                })
```

**Comment by dankelley** [comment]
Actually, there would be no problem with the `subset()` if I just made our new function, `ctdFindProfilesRBR()`, save *a single* lon-lat location in the metadata.  And, frankly, that makes sense to me.  The lon and lat are **not** measured by the instrument, so that says they ought to go into `metadata` (to fit a general oce convention).  But CTD objects usually have just a single location attached to them.

It seems that there are 3 choices:
1. Make `ctdFindProfileRBR()` save just a single lon-lat value, in the `metadata` slot. (This is my preferred approach.)
2. Put lon and lat into the `data` slot.
3. Make `subset,ctd-method()` do special things with some elements of `metadata`.  (I reject this, for the reasons above.)

**Comment by dankelley** [comment]
With the commit as below, I am rejecting my proposal as stated in the title.  This closes the issue.

commit 006598fbf7588e2b9d576a749a2b2a17d570598b (HEAD -> rsk, origin/rsk)
Author: dankelley <kelley.dan@gmail.com>
Date:   Fri Dec 30 14:08:18 2022 -0400

    ctdFindProfilesRBR() stores 1 lon-lat per profile

**Comment by dankelley** [comment]
The "rsk" branch has been merged into the "develop" branch (commit c3bb3fd3e517cc2cd94fbd288ca1767a2a1ec839) and it is the latter that ought to be used for further tests.  

If large-scale recoding is required to address any issues, I'll revive "rsk" but for now, I ask @Brendan-Higgins and @richardsc (the two most involved in the RSK work over the past few weeks) to focus on "develop".  Also, I ask that they create new issues with narrow titles, if new problems arise or new features are desired.  Issue drift gets confusing on github, owing to the lack of threaded comments.

Thanks for the patience on this.

Dan.
